apache2/waf_lock/waf_lock.cpp

```diff
@@ -0,0 +1,533 @@
+#include"waf_lock_external.h"
+#include"waf_lock_internal.h"
+
+
+/**
+**========================================================
+** Waf Lock Internal Function
+**========================================================
+*/
+
+/**
+** Create and initialize a waf_lock, if the lock with given key exists, just link to the lock.
+** The new_lock should have been created before calling this function.
+** @param new_lock:        The pointer to save the information of the lock.
+** @param lock_name:       The name of lock.
+** @param lock_name_length:The length of the lock_name.
+** return: WAF_SUCCESS_LOCK_CREATE if successfully created a new lock,
+**         WAF_SUCCESS_LOCK_OPEN if successfully link to an existed lock,
+**         WAF_LOCK_ERROR if failed.
+*/
+int lock_create(struct waf_lock *new_lock, const char* lock_name, int lock_name_length, const char* owner) {
+#ifndef _WIN32
+    union semun sem_union;
+    int lock_id = -1;
+    uid_t uid;
+    gid_t gid;
+#else
+    int read_lock_name_len = 0;
+    int write_lock_name_len = 0;
+    char* read_lock_name = NULL;
+    char* write_lock_name = NULL;
+    bool lock_exists = false;
+#endif
+    if (new_lock == NULL)
+        return WAF_LOCK_ERROR_HANDLE_NULL;
+#ifndef _WIN32
+    lock_id = Waf_lock_hash(lock_name, lock_name_length, DEFAULT_WAF_LOCK_ID_RANGE);
```

Comment:
I think we just need to do string and ID match. Do we need the murmur hash?

Reply:
will let user input the id